app-libs/sgx-runtime/pallets/credits/src/lib.rs

```diff
@@ -0,0 +1,415 @@
+		TooManyEntries,
+		Unauthorized,
+		ClassAdminUndefined,
+		ExpiryInPast,
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn credits)]
+	pub type Credits<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		CreditClassId,
+		Blake2_128Concat,
+		T::AccountId,
+		SortedCreditsStore<BalanceOf<T>, T::Moment>,
+		ValueQuery,
+	>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn admin)]
+	pub(super) type Admin<T: Config> =
+		StorageMap<_, Blake2_128Concat, CreditClassId, T::AccountId, OptionQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn total_minted_by)]
+	pub type TotalMintedBy<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		CreditClassId,
+		Blake2_128Concat,
+		T::AccountId,
+		BalanceOf<T>,
+		ValueQuery,
+	>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn total_redeemed_by)]
+	pub type TotalRedeemedBy<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		CreditClassId,
+		Blake2_128Concat,
+		T::AccountId,
+		BalanceOf<T>,
+		ValueQuery,
+	>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn total_deposit)]
+	pub type TotalDeposit<T: Config> =
+		StorageMap<_, Blake2_128Concat, CreditClassId, BalanceOf<T>, ValueQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		sp_core::H256: From<<T as frame_system::Config>::Hash>,
+	{
+		/// create a new credit class
+		#[pallet::call_index(0)]
+		#[pallet::weight((<T as Config>::WeightInfo::create_class(), DispatchClass::Normal, Pays::Yes)
+        )]
+		pub fn create_class(
+			origin: OriginFor<T>,
+			id: CreditClassId,
+			deposit: BalanceOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let sender = ensure_signed(origin)?;
+			ensure!(!<Credits<T>>::contains_prefix(id), Error::<T>::ClassIdExists);
+			T::Currency::reserve(&sender, deposit)?;
+			TotalDeposit::<T>::insert(id, deposit);
+			<Credits<T>>::insert(id, &sender, SortedCreditsStore::new());
+			<Admin<T>>::insert(id, &sender);
+			Self::deposit_event(Event::CreatedClass { id });
+			Ok(().into())
+		}
+
+		/// create a new credit class
+		#[pallet::call_index(1)]
+		#[pallet::weight((<T as Config>::WeightInfo::create_class(), DispatchClass::Normal, Pays::Yes)
+        )]
+		pub fn destroy_class(
+			origin: OriginFor<T>,
+			id: CreditClassId,
+		) -> DispatchResultWithPostInfo {
+			let sender = ensure_signed(origin)?;
+			ensure!(<Credits<T>>::contains_prefix(id), Error::<T>::InvalidClassId);
+			let admin = Self::admin(id).ok_or(Error::<T>::ClassAdminUndefined)?;
+			ensure!(admin == sender, Error::<T>::Unauthorized);
+			// TODO: consider limiting to avoid overweight execution
+			let _ = Credits::<T>::clear_prefix(id, u32::max_value(), None);
+			let unreserve = Self::total_deposit(id);
+			TotalDeposit::<T>::remove(id);
+			Admin::<T>::remove(id);
+			TotalMintedBy::<T>::remove(id, &admin);
+			TotalRedeemedBy::<T>::remove(id, &admin);
+			T::Currency::unreserve(&admin, unreserve);
+			Self::deposit_event(Event::DestroyedClass { id });
+			Ok(().into())
+		}
+
+		/// create a new credit class
```

Comment:
bad doc. After trying to understand your code, I would write


```suggestion
		/// Claim credits belonging to the commitment accout corresponding to the `secret`.
```